File: projects/relay/engine-java/src/main/java/com/llmeval/relay/engine/Track.java
package com.llmeval.relay.engine;

import java.util.ArrayList;
import java.util.List;

/**
 * Stage generation and answer checking.
 *
 * <p>Every sentence built here is a corpus byte: the whole track rides inside
 * {@code track_generated}, so the conformance digest covers it. This class must agree with
 * {@code relay_engine/track.py} to the space — including the space that joins sentences and the
 * question mark that ends them.
 *
 * <p>The tier never reaches the prose. A generator that made hard stages read as harder would
 * hand the runners the one thing the game asks them to judge, and it would pass every test that
 * only checked answers.
 */
public final class Track {

    private Track() {}

    public static final List<String> FAMILIES = List.of("chain", "cipher", "order");

    public static final int TRACK_STAGES = 10;
    public static final int[] TIER_MULTISET = {1, 1, 1, 1, 2, 2, 2, 2, 3, 3};

    static final int[] CHAIN_STEPS = {0, 2, 4, 6};
    static final List<String> CIPHER_SHORT =
            List.of("iron", "moss", "lamp", "reed", "sail", "vine", "clay", "dusk");
    static final List<String> CIPHER_LONG =
            List.of("mariner", "lantern", "kestrel", "harvest", "tundra", "cobalt",
                    "quarry", "silence");
    static final List<String> ORDER_NAMES =
            List.of("ada", "brun", "cyd", "dev", "esme", "fen", "gil", "hana");
    static final int[] ORDER_ITEMS = {0, 3, 4, 5};
    static final String[] ORDINALS = {"first", "second", "third", "fourth", "fifth"};

    /** How an answer is compared: forgiving about wrapping, strict about the token. */
    public static String normalise(String answer) {
        String trimmed = answer.strip();
        while (trimmed.endsWith(".")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        return trimmed.strip().toLowerCase();
    }

    public static List<Stage> generate(Rng rng) {
        return generate(rng, TRACK_STAGES);
    }

    /** Draw order is spec: tiers first (one shuffle), then each stage's family and body. */
    public static List<Stage> generate(Rng rng, int stages) {
        List<Integer> tiers = new ArrayList<>();
        for (int i = 0; i < stages; i++) {
            tiers.add(TIER_MULTISET[i]);
        }
        rng.shuffle(tiers);

        List<Stage> built = new ArrayList<>();
        for (int index = 0; index < stages; index++) {
            String family = FAMILIES.get(rng.below(FAMILIES.size()));
            int tier = tiers.get(index);
            String[] body = switch (family) {
                case "chain" -> chain(rng, tier);
                case "cipher" -> cipher(rng, tier);
                default -> order(rng, tier);
            };
            built.add(new Stage(String.format("stage-%02d", index + 1), family, tier,
                    body[0], body[1]));
        }
        return built;
    }

    // -- families --------------------------------------------------------

    private static String[] chain(Rng rng, int tier) {
        int value = rng.between(1, 20);
        List<String> parts = new ArrayList<>();
        parts.add("Start with " + value + ".");
        // Multiplying by three only appears at tier 3.
        int kinds = tier < 3 ? 3 : 4;
        for (int step = 0; step < CHAIN_STEPS[tier]; step++) {
            int kind = rng.below(kinds);
            if (kind == 0) {
                int n = rng.between(2, 10);
                value += n;
                parts.add("Add " + n + ".");
            } else if (kind == 1) {
                int n = rng.between(2, 10);
                value -= n;
                parts.add("Subtract " + n + ".");
            } else if (kind == 2) {
                value *= 2;
                parts.add("Double it.");
            } else {
                value *= 3;
                parts.add("Triple it.");
            }
        }
        parts.add("What number do you end with?");
        return new String[] {String.join(" ", parts), Integer.toString(value)};
    }

    private static String[] cipher(Rng rng, int tier) {
        List<String> pool = tier == 1 ? CIPHER_SHORT : CIPHER_LONG;
        String word = pool.get(rng.below(pool.size()));
        int shift = rng.between(1, 25);
        String encoded = caesar(word, shift).toUpperCase();

        String prompt;
        if (tier < 3) {
            prompt = "Every letter of a word was shifted forward " + shift
                    + " places through the alphabet, wrapping from z back to a, giving "
                    + encoded + ". What was the original word?";
        } else {
            prompt = "Every letter of a word was shifted forward through the alphabet by the "
                    + "same unknown number of places, wrapping from z back to a, giving "
                    + encoded + ". The original word begins with '" + word.charAt(0)
                    + "'. What was the original word?";
        }
        return new String[] {prompt, word};
    }

    static String caesar(String word, int shift) {
        StringBuilder out = new StringBuilder();
        for (int i = 0; i < word.length(); i++) {
            out.append((char) ('a' + (word.charAt(i) - 'a' + shift) % 26));
        }
        return out.toString();
    }

    private static String[] order(Rng rng, int tier) {
        int count = ORDER_ITEMS[tier];
        List<String> order = rng.sample(ORDER_NAMES, count);

        List<String> facts = new ArrayList<>();
        for (int i = 0; i < count - 1; i++) {
            facts.add(order.get(i) + " is somewhere before " + order.get(i + 1) + ".");
        }
        if (tier == 3) {
            // True, and it rules out nothing the chain did not already rule out.
            facts.add(order.get(rng.below(count - 1)) + " is not last.");
        }
        rng.shuffle(facts);

        int position = rng.below(count);
        String prompt = count + " runners crossed the line one at a time. "
                + String.join(" ", facts)
                + " Who finished " + ORDINALS[position] + "? Answer with one name.";
        return new String[] {prompt, order.get(position)};
    }
}

File: projects/relay/engine-python/src/relay_engine/track.py
"""The track: stage generation and answer checking.

Every stage is generated forward from the seed, so the answer is known by
construction and never has to be solved by the engine. Three families with a
difficulty ladder, per game-rules.md.

Two properties are load-bearing and easy to break:

- **The prompts are corpus bytes.** They ride inside the transcript, so the
  conformance digest covers them: a Java port that renders one space
  differently fails every vector. Everything here is ASCII, integer-only, and
  built by plain concatenation — no locale-sensitive formatting anywhere.
- **The tier is not in the prompt.** It is metadata the engine keeps and the
  transcript reveals only at `game_ended`. A generator that leaked difficulty
  into the wording (longer sentences for tier 3, say) would quietly hand the
  runners the one thing the game asks them to judge for themselves.
"""

from __future__ import annotations

from dataclasses import dataclass

from .rng import Rng

FAMILIES = ("chain", "cipher", "order")

#: Stages on the track, and the tier multiset dealt across them. The multiset
#: is shuffled, so position tells a runner nothing.
TRACK_STAGES = 10
TIER_MULTISET = (1, 1, 1, 1, 2, 2, 2, 2, 3, 3)

#: chain: how many operations per tier.
CHAIN_STEPS = {1: 2, 2: 4, 3: 6}
#: cipher: which word pool per tier. Tier 3 reuses the long pool and withholds
#: the shift instead — harder by inference, not by length.
CIPHER_SHORT = ("iron", "moss", "lamp", "reed", "sail", "vine", "clay", "dusk")
CIPHER_LONG = ("mariner", "lantern", "kestrel", "harvest", "tundra", "cobalt",
               "quarry", "silence")
#: order: names are single lowercase ASCII words so the answer is one token.
ORDER_NAMES = ("ada", "brun", "cyd", "dev", "esme", "fen", "gil", "hana")
ORDER_ITEMS = {1: 3, 2: 4, 3: 5}
ORDINALS = ("first", "second", "third", "fourth", "fifth")


@dataclass(frozen=True)
class PublicStage:
    """What a runner is allowed to see: the puzzle, and nothing else."""

    id: str
    family: str
    prompt: str


@dataclass(frozen=True)
class Stage:
    id: str
    family: str
    tier: int
    prompt: str
    answer: str

    def public(self) -> PublicStage:
        return PublicStage(self.id, self.family, self.prompt)


def normalise(answer: str) -> str:
    """How an answer is compared. Forgiving about wrapping, strict about the
    token itself — models add trailing full stops and capitals, and neither is
    a wrong answer."""
    return answer.strip().rstrip(".").strip().lower()


def generate(rng: Rng, stages: int = TRACK_STAGES) -> tuple[Stage, ...]:
    """Draw order is spec: tiers first (one shuffle), then each stage's family
    and body in track order."""
    tiers = list(TIER_MULTISET[:stages])
    rng.shuffle(tiers)

    built: list[Stage] = []
    for index in range(stages):
        family = FAMILIES[rng.below(len(FAMILIES))]
        tier = tiers[index]
        prompt, answer = _BUILDERS[family](rng, tier)
        built.append(Stage(f"stage-{index + 1:02d}", family, tier, prompt, answer))
    return tuple(built)


# -- families --------------------------------------------------------------


def _chain(rng: Rng, tier: int) -> tuple[str, str]:
    """Integer arithmetic, applied in order. Mechanical: a program is perfect
    at these, and so is a small model — which is exactly why the track needs
    the other two families."""
    value = rng.between(1, 20)
    parts = [f"Start with {value}."]
    # Multiplying by three only appears at tier 3, where the numbers also get
    # large enough that a careless step compounds.
    kinds = 3 if tier < 3 else 4
    for _ in range(CHAIN_STEPS[tier]):
        kind = rng.below(kinds)
        if kind == 0:
            n = rng.between(2, 10)
            value += n
            parts.append(f"Add {n}.")
        elif kind == 1:
            n = rng.between(2, 10)
            value -= n
            parts.append(f"Subtract {n}.")
        elif kind == 2:
            value *= 2
            parts.append("Double it.")
        else:
            value *= 3
            parts.append("Triple it.")
    parts.append("What number do you end with?")
    return " ".join(parts), str(value)


def _cipher(rng: Rng, tier: int) -> tuple[str, str]:
    """A Caesar shift. Tier 3 withholds the shift and gives a crib instead, so
    the puzzle becomes two steps: recover the shift, then apply it."""
    pool = CIPHER_SHORT if tier == 1 else CIPHER_LONG
    word = pool[rng.below(len(pool))]
    shift = rng.between(1, 25)
    encoded = _caesar(word, shift).upper()

    if tier < 3:
        prompt = (f"Every letter of a word was shifted forward {shift} places "
                  f"through the alphabet, wrapping from z back to a, giving "
                  f"{encoded}. What was the original word?")
    else:
        prompt = (f"Every letter of a word was shifted forward through the "
                  f"alphabet by the same unknown number of places, wrapping "
                  f"from z back to a, giving {encoded}. The original word "
                  f"begins with '{word[0]}'. What was the original word?")
    return prompt, word


def _caesar(word: str, shift: int) -> str:
    out = []
    for ch in word:
        out.append(chr(ord("a") + (ord(ch) - ord("a") + shift) % 26))
    return "".join(out)


def _order(rng: Rng, tier: int) -> tuple[str, str]:
    """Pairwise constraints that pin one total order. The constraints are
    always the n-1 consecutive links — enough for exactly one solution — with
    their statement order shuffled so the reader has to reassemble the chain.
    Tier 3 adds one true-but-redundant negative fact: noise, not ambiguity."""
    count = ORDER_ITEMS[tier]
    order = rng.sample(list(ORDER_NAMES), count)

    facts = [f"{order[i]} is somewhere before {order[i + 1]}."
             for i in range(count - 1)]
    if tier == 3:
        # Any name except the last one; the statement is true, and it rules out
        # nothing the chain did not already rule out.
        not_last = order[rng.below(count - 1)]
        facts.append(f"{not_last} is not last.")
    rng.shuffle(facts)

    position = rng.below(count)
    prompt = (f"{count} runners crossed the line one at a time. "
              + " ".join(facts)
              + f" Who finished {ORDINALS[position]}? "
                f"Answer with one name.")
    return prompt, order[position]


_BUILDERS = {"chain": _chain, "cipher": _cipher, "order": _order}
